src/client/lazy-app/Modal/index.tsx

```diff
@@ -0,0 +1,128 @@
+import { h, Component, VNode, Fragment } from 'preact';
+import * as style from './style.css';
+import 'add-css:./style.css';
+import { linkRef } from 'shared/prerendered-app/util';
+
+interface Props {}
+
+export interface ModalMessage {
+  icon: VNode;
+  title: string;
+  content: VNode;
+}
+
+interface State {
+  message: ModalMessage;
+  shown: boolean;
+}
+
+export default class Modal extends Component<Props, State> {
+  state: State = {
+    message: {
+      icon: <svg></svg>,
+      title: 'default',
+      content: <Fragment></Fragment>,
+    },
+    shown: false,
+  };
+
+  private dialogElement!: HTMLDialogElement;
+  static modalInstance?: Modal | undefined;
+
+  componentDidMount() {
+    // Once a transition ends, check if the modal should be closed (not just hidden)
+    // dialog.close() instantly hides the modal, so we call it AFTER fading it out i.e. on transition end
+    this.dialogElement.addEventListener(
+      'transitionend',
+      this._closeOnTransitionEnd.bind(this),
+    );
+    this.dialogElement.setAttribute('inert', 'enabled');
+
+    Modal.modalInstance = this;
+  }
+
+  private _closeOnTransitionEnd() {
+    // If modal does not exist
+    // Or if it's not being closed at the moment
+    if (
+      !this.dialogElement ||
+      !this.dialogElement.classList.contains(style.modalClosing)
+    )
+      return;
+
+    this.dialogElement.close();
+    this.dialogElement.classList.remove(style.modalClosing);
+    this.dialogElement.setAttribute('inert', 'enabled');
+  }
+
+  static showModal(message: ModalMessage) {
+    Modal.modalInstance?._showModal(message);
+  }
+
+  static hideModal() {
+    Modal.modalInstance?._hideModal();
+  }
+
+  private _showModal(message: ModalMessage) {
+    if (!this.dialogElement) return;
+
+    this.setState({
+      message: message,
+      shown: true,
+    });
+
+    // Actually show the modal
+    this.dialogElement.removeAttribute('inert');
+    this.dialogElement.showModal();
+  }
+
+  private _hideModal() {
+    if (!this.dialogElement || !this.dialogElement.open) return;
+
+    // Make the modal fade out
+    this.dialogElement.classList.add(style.modalClosing);
+
+    this.setState({
+      message: { ...this.state.message },
+      shown: false,
+    });
+  }
+
+  private _onKeyDown(e: KeyboardEvent) {
+    // Default behaviour of <dialog> closes it instantly when you press Esc
+    // So we hijack it to smoothly hide the modal
+    if (e.key === 'Escape' || e.keyCode == 27) {
```

Comment:
I [don’t think](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent#:~:text=Toggle%20history-,key,-51) we need both.

```suggestion
    if (e.key === 'Escape') {
```